File: leetcode/dp/Climbing_Stairs/java/SolutionTabulation.java
// https://leetcode.com/problems/climbing-stairs/description/
// 70. Climbing Stairs

// Tabulation solution | Bottom-Up | Dynamic Programming
// Time Complexity: O(n)
// Space Complexity: O(n)
class SolutionTabulation {
    public int climbStairs(int n) {
        // tabulation array to store the results of the subproblems
        int ways[] = new int[n+1];
        // base cases
        ways[1] = 1;
        ways[2] = 2;
        for (int i=3; i<n+1; i++) {
            ways[i] = ways[i-2] + ways[i-1];
        }
        return ways[n];
    }
    public static void main(String[] args) {
        int n = 45, expected = 1836311903;
        System.out.println("Solution Tabulation");
        SolutionTabulation soltab = new SolutionTabulation();
        int actual = soltab.climbStairs(n);
        System.out.println(actual);
        
        if (actual == expected) {
            System.out.println("Test Passed");
        } else {
            System.out.println("Test Failed");
        }
    }
}

File: leetcode/dp/Climbing_Stairs/python/SolutionTabulation.py
# https://leetcode.com/problems/climbing-stairs/
# 70. Climbing Stairs

# Tabulation Solution | Bottom-Up | Dynamic Programming
# Time Complexity: O(n) because we are iterating through the array once
# Space Complexity: O(n) because we are using an array to store the results of the subproblems
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [0] * (n+1)
        # base cases
        dp[1] = 1 # first stair: can climb in only one way
        dp[2] = 2 # sec stair: 2 ways: (1+1) or 2 at a time
        # done upto index 2, star wil index 3
        for i in range(3, n+1):
            dp[i] = dp[i-2] + dp[i-1]
        return dp[n]

if __name__ == "__main__":
    n = 45
    expected = 1836311903
    solution = Solution()
    actual = solution.climbStairs(n)
    if actual == expected:
        print("Test Passed")
    else:
        print(f"Test Failed: Expected {expected}, Actual {actual}")